ouroboros-consensus/docs/VersioningSchemeDecision.md

```diff
@@ -0,0 +1,465 @@
+# Choosing a Versioning Schemes
+
+This document records our discussions around choosing how to version our packages.
+
+In its current state, the document is primarily focused on one question: for each individual package, how should the version listed in the `.cabal` file on the main integration branch relate to the version in the `.cabal` file in the latest release?
+In particular, I think this question is almost entirely independent of usings `git` branches versus `git` tags for releases, SemVer versus PVP, etc.
+This document should eventually specify that level of detail as well.
+
+Given its current focus, without loss of generality, suppose the repository contains only one package while considering the proposals below.
+
+## Desiderata
+
+- *Conventional Version Numbers*.
+  We'd like our release versions to have the standard shape of `A.B.C`, and the standard rules that `A` must increase for breaking changes, `B` must increase for backwards-compatible changes, and `C` must increase for anything else (eg bugfixes/documentation/etc aka "patches").
+
+    - Some of our proposals below don't care about the distinction between `A` and `B`.
+      Instead of repeating the semantics of `A` and `B` in each proposal, we refer to `next(A.B)` with the intention that the developer's determine whether to increment `A` or `B`.
+
+    - You are of course free to decide any PR should increment `A` or `B` even if the proposed rules do not require it.
+      Ultimately, the only risk of such spurious increases is confusing/inconveniencing downstream users.
+
+    - Note, in the context of [the PVP](https://pvp.haskell.org/) eg, our `A` dimension here would itself denote a pair.
+
+- *Simplicity and Familiarity*.
+  We'd like the versioning scheme to be simple to explain and ideally already well-established.
+
+- *Ease of Execution*.
+  We'd like the process of cutting a release to merely involve following a very simple checklist.
+  We'd like it to require as few inputs, discussions, decisions, etc as possible.
+
+- *Distinguished Development Versions*.
```

Comment:
I had a somewhat evil idea. 

```
flag allow-dev-version
   description: Allow the use of development versions of this package.
   manual: True
   default: false
```

Then when you are on a development version you have:
```
if !flag(allow-dev-version)
  buildable: False
```
... and not when you do a release. You can set it locally in your `cabal.project` so you can always use the versions, but downstream won't be able to use them without explicitly setting `allow-dev-version`.

So the release process is something like:
- Make a commit C1 which bumps the version numbers 
- Make a commit C2 which removes all the `allow-dev-version` guards
- Tag that commit as the release
- Revert C2

Reply:
(I also want to register that I'm less convinced than you that this is a big problem)

Reply:
I would not be opposed to this. I think it is a nice use of the `buildable` flag and can be made consistent in the whole project. Configure it once, be safe forever :smile: